### PR TITLE
webgpu: Optimize getSamplerAtOutputCoords

### PR DESCRIPTION
**Pull request by qjia7**
PERF

~160% perf improvement is observed for the total time of Relu6 in MobieNetV1 model.

To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5105)

### CODE REVIEW

**Comment by qjia7** [comment]
@axinging @haoyunfeix @xhcao Please take a look, thanks.

File: tfjs-backend-webgpu/src/shader_preprocessor.ts [hunk 1]
```diff
@@ -357,6 +362,35 @@ function getSamplerAtOutputCoords(
   const outRank = outShape.length;
   const type = getCoordsDataType(outRank);
 
+  // If the inShape equals the outShape and the dispatch layout is flat, we can
+  // directly use |gl_GlobalInvocationID.x| as the index and don't need coords
+  // conversion between these two shapes.
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/shader_preprocessor.ts:367]: Previously `getAAtOutCoords()` looks like below:
```
  int getFlatIndex(ivec4 coords, ivec4 shape) {
    return int(dot(coords, ivec4(
      shape.y * shape.z * shape.w, shape.z * shape.w, shape.w, 1.)));
  }

   ivec3 getCoordsFromFlatIndex(int index) {
      int d0 = index / outShapeStrides[0]; index -= d0 * outShapeStrides[0];int d1 = index / outShapeStrides[1]; int d2 = index - d1 * outShapeStrides[1];
      return ivec3(d0,d1,d2);
    }
  
    ivec3 getOutputCoords() {
      return getCoordsFromFlatIndex(int(gl_GlobalInvocationID.x));
    }

    float getAAtOutCoords() {
      ivec3 coords = getOutputCoords();
      
      return float(A[getFlatIndex(ivec3(coords[0], coords[1], coords[2]), aShape)]);
    }
```
Now, `getAAtOutCoords()` is simplified as below. It saves many calculation in shader. This is very useful for unary ops and non-broadcast binary ops.
```
      float getAAtOutCoords() {
        return float(A[gl_GlobalInvocationID.x]);
      }

**Review by qjia7 — COMMENTED** [review]
@kainino0x @lina128 @jinjingforever Take a look, thanks.

**Review by lina128 — COMMENTED** [review]
LGTM. Please have Kai review it too.

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5105#-:-Maawkhq5wIcDqctWQid:bfq41t)* status: 0 of 1 approvals obtained (waiting on @jinjingforever and @kainino0x)

**Review by kainino0x — APPROVED** [review]
Nice work!

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5105#-:-Maf1MLU9JI76fTMLEej:bdtx8r6)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @jinjingforever and @qjia7)

___
*[tfjs-backend-webgpu/src/backend_webgpu.ts, line 698 at r1](https://reviewable.io/reviews/tensorflow/tfjs/5105#-Maf0vg341-NtVQ3Lnaw:-Maf0vg4-ZF7idGuT2oN:b-jqf97m) ([raw file](https://github.com/tensorflow/tfjs/blob/a558577d4011303466857be42b14f59bfe872e9c/tfjs-backend-webgpu/src/backend_webgpu.ts#L698)):*
> ```XML
>         d => backend_util.getBroadcastDims(d.shape, output.shape));
>     const inputShapesEqualsOutShape =
>         inputsData.map(d => util.arraysEqual(d.shape, output.shape)).join('_');
> ```

This looks like it would produce something like true_true_true_true, is that necessary? Maybe just `inputsData.every(d => util.arraysEqual(d.shape, output.shape))` instead?

**Comment by qjia7** [comment]
> >         d => backend_util.getBroadcastDims(d.shape, output.shape));
> >     const inputShapesEqualsOutShape =
> >         inputsData.map(d => util.arraysEqual(d.shape, output.shape)).join('_');
> > ```
> 
> This looks like it would produce something like true_true_true_true, is that necessary? Maybe just `inputsData.every(d => util.arraysEqual(d.shape, output.shape))` instead?

true_true_true_true is needed. For example, we have two inputs A and B. The result may be true_true or true_false or false_true, or false_false. Each of them will correspond to different shader.
The shader snippet is like below:
true_true
```
      float getAAtOutCoords() {
        return float(A[gl_GlobalInvocationID.x]);
      }

      float getBAtOutCoords() {
        return float(B[gl_GlobalInvocationID.x]);
      }
```
true_false
```
      float getAAtOutCoords() {
        return float(A[gl_GlobalInvocationID.x]);
      }

      float getBAtOutCoords() {
        ivec3 coords = getOutputCoords();
      
        return float(B[getFlatIndex(ivec3(coords[0], coords[1], coords[2]), aShape)]);
      }
```

**Comment by kainino0x** [comment]
Ahh, I should have realized that, thanks for the explanation. LGTM

**Pull request merged**